### PR TITLE
gh-122581: Avoid data races when collecting parser statistics

### PR DESCRIPTION
**Pull request by lysnikolaou**
* Issue: gh-122581

### CODE REVIEW

**Review by pablogsal — APPROVED** [review]

File: Parser/pegen.c [hunk 1]
```diff
@@ -311,6 +321,8 @@ _PyPegen_get_memo_statistics(void)
     if (ret == NULL) {
         return NULL;
     }
+
+    MUTEX_LOCK();
     for (int i = 0; i < NSTATISTICS; i++) {
         PyObject *value = PyLong_FromLong(memo_statistics[i]);
         if (value == NULL) {
```
Comment (kumaraditya303) [inline Parser/pegen.c:328]: You need unlock here `if value == NULL`

File: Parser/pegen.c [hunk 2]
```diff
@@ -323,6 +335,7 @@ _PyPegen_get_memo_statistics(void)
             return NULL;
```
Comment (kumaraditya303) [inline Parser/pegen.c:337]: Same

File: Parser/pegen.c (hunk 1 above)
Comment (lysnikolaou) [reply Parser/pegen.c:328]: Thanks for the review! Addressed.

**Review by kumaraditya303 — APPROVED** [review]
LGTM

**Pull request merged**